pkg/nfd-master/nfd-master.go

```diff
@@ -440,10 +451,110 @@ func (m *nfdMaster) SetLabels(c context.Context, r *pb.SetLabelsRequest) (*pb.Se
 			klog.Errorf("failed to advertise labels: %v", err)
 			return &pb.SetLabelsReply{}, err
 		}
+
+		// set taints
+		var taints []corev1.Taint
+		if m.args.EnableTaints {
+			taints = crTaints
+		}
+
+		// Call setTaints even though the feature flag is disabled. This
+		// ensures that we delete NFD owned stale taints when flag got
+		// turned off.
+		err = m.setTaints(cli, taints, r.NodeName)
+		if err != nil {
+			return &pb.SetLabelsReply{}, err
+		}
 	}
 	return &pb.SetLabelsReply{}, nil
 }
 
+// setTaints sets node taints and annotations based on the taints passed via
+// nodeFeatureRule custom resorce. If empty list of taints is passed, currently
+// NFD owned taints and annotations are removed from the node.
+func (m *nfdMaster) setTaints(cli *kubernetes.Clientset, taints []corev1.Taint, nodeName string) error {
+	// Fetch the node object.
+	node, err := m.apihelper.GetNode(cli, nodeName)
+	if err != nil {
+		return err
+	}
+
+	// De-serialize the taints annotation into Taint type for comparision below.
+	annotations := node.GetAnnotations()
+	// Keep the backup of old annotation.
+	oldAnnotation := make(map[string]string)
+	for key, value := range annotations {
+		oldAnnotation[key] = value
+	}
+	oldTaints := []corev1.Taint{}
+	if val, ok := annotations[nfdv1alpha1.NodeTaintsAnnotation]; ok {
+		sts := strings.Split(val, ",")
```

Comment:
... and then we can simplify this pme

```suggestion
	if oldAnnotation != "" {
		sts := strings.Split(oldAnnotation, ",")
```

Reply:
done